### ISSUE TITLE
Parameter matching fails on keyword arguments

### ISSUE THREAD

**Issue by CodingAnarchy** [freerange/mocha#647]
Pretty basic but surprising bug that I ran into just now, where parameter matching on `expect(...).with(...)` doesn't work properly with keyword arguments.

```ruby
        class Foo
          def self.bar(a:, b:, c:)
            a + b + c
          end
        end

        def test_mocha_mocking_kw_args
          mock = mock("foo")
          mock.expects(:bar).with(a: 1, b: 2, c: is_a(Integer)).returns(6)

          mock.bar(a: 1, b: 2, c: 3)
        end
```

This results in the following error:

```
unexpected invocation: #<Mock:foo>.bar(:a => 1, :b => 2, :c => 3)
unsatisfied expectations:
- expected exactly once, invoked never: #<Mock:foo>.bar(:a => 1, :b => 2, :c => is_a(Integer))
```

The positional argument equivalent works just fine:

```ruby
        class Foo
          def self.bar(a, b, c)
            a + b + c
          end
        end

        def test_mocha_mocking_positional_args
          mock = mock("foo")
          mock.expects(:bar).with(1, 2, is_a(Integer)).returns(6)

          mock.bar(1, 2, 3)
        end
```

Mocha: v2.2.0
Ruby: 3.3.1

**Comment by floehopper** [comment]
@CodingAnarchy 

Thanks for reporting this. It looks as if you've made a mistake in your examples - I imagine you intended to mock & call `Foo.bar` from the testx, but instead you are mocking and calling `Mock#bar`. Anyway, that doesn't take away from the fact that there seems to be an issue with the nested `is_a` parameter matcher - "nested", because I believe the kwargs are treated as a `Hash` internally. I've tried to more clearly illustrate the problem below - let me know if I've missed the mark:

```ruby
require "bundler/setup"
require "minitest/autorun"
require "mocha/minitest"
require "mocha/mock"

class FooTest < Minitest::Test
  def setup
    @mock = mock("foo")
  end

  def test_mocking_positional_args
    @mock.expects(:bar).with(1, 2, is_a(Integer))

    @mock.bar(1, 2, 3)
  end

  def test_mocking_positional_args_with_nested_matcher
    @mock.expects(:bar).with(1, 2, is_a(Integer))

    @mock.bar(1, 2, 3)
  end

  def test_mocha_mocking_kw_args
    @mock.expects(:bar).with(a: 1, b: 2, c: 3)

    @mock.bar(a: 1, b: 2, c: 3)
  end

  def test_mocha_mocking_kw_args_with_nested_matcher
    @mock.expects(:bar).with(a: 1, b: 2, c: is_a(Integer))

    @mock.bar(a: 1, b: 2, c: 3)
    # unexpected invocation: #<Mock:foo>.bar(:a => 1, :b => 2, :c => 3)
    # unsatisfied expectations:
    # - expected exactly once, invoked never: #<Mock:foo>.bar(:a => 1, :b => 2, :c => is_a(Integer))
  end
end
```

**Comment by CodingAnarchy** [comment]
@floehopper Yeah, I meant to mock `Foo#bar` there. Thanks for catching that.

Your example does demonstrate the issue I ran into though, so I think you've got it spot on in the replication.

**Comment by floehopper** [comment]
I just opened this PR to fix this: https://github.com/freerange/mocha/pull/648

**Comment by floehopper** [comment]
Released in [v2.3.0](https://rubygems.org/gems/mocha/versions/2.3.0).
